Fix ProcessGroupNCCL profiling when profiler is not run with use_cuda

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49072 Test distributed collectives profiling with Gloo on GPU
* **#48946 Fix ProcessGroupNCCL profiling when profiler is not run with use_cuda**

Move `recordFunctionEndCallback` to after the blocking portion of launching the NCCL kernel, and remove `addCallback` since it runs the lambda inline anyways, and triggers unnecessary CUDA stream logic. If we want CUDA operations such as NCCL kernels accurately profiled, we should use the profiler with `use_cuda=True`. However, we are currently debugging a deadlock for the `use_cuda=True` case, fix is being tracked in https://github.com/pytorch/pytorch/issues/48987. 

To ensure that the tests are no longer flaky, submitted this PR to ci-all: https://github.com/pytorch/pytorch/pull/48947 and ran the test a bunch of times ssh'd into the CI machine and made sure they don't fail.

Differential Revision: [D25368322](https://our.internmc.facebook.com/intern/diff/D25368322/)

rohan-varma committed:
[wip][not for review] Fix ProcessGroupNCCL profiling when profiler is not run with use_cuda

Still does not work with use_cuda=True. Debugging the deadlock
currently.

Differential Revision: [D25368322](https://our.internmc.facebook.com/intern/diff/D25368322/)

[ghstack-poisoned]

mrzzd commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1126,6 +1115,17 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
   work->opTimeout_ = opTimeout_;
   work->store_ = store_;
 
+  if (work->recordFunctionEndCallback_) {
```
Thanks for working on this @rohan-varma . Would you elaborate why it matters to add callback later here?

rohan-varma replied:
@mrzzd As discussed offline, it's equivalent to directly invoking the callback inline since the callback doesn't do any GPU work. See https://github.com/pytorch/pytorch/pull/48561 for more of a discussion/context.

rohan-varma committed:
Update on "Fix ProcessGroupNCCL profiling when profiler is not run with use_cuda"


Move `recordFunctionEndCallback` to after the blocking portion of launching the NCCL kernel, and remove `addCallback` since it runs the lambda inline anyways, and triggers unnecessary CUDA stream logic. If we want CUDA operations such as NCCL kernels accurately profiled, we should use the profiler with `use_cuda=True`. However, we are currently debugging a deadlock for the `use_cuda=True` case, fix is being tracked in https://github.com/pytorch/pytorch/issues/48987. 

To ensure that the tests are no longer flaky, submitted this PR to ci-all: https://github.com/pytorch/pytorch/pull/48947 and ran the test a bunch of times ssh'd into the CI machine and made sure they don't fail.

Differential Revision: [D25368322](https://our.internmc.facebook.com/intern/diff/D25368322/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Fix ProcessGroupNCCL profiling when profiler is not run with use_cuda"


Move `recordFunctionEndCallback` to after the blocking portion of launching the NCCL kernel, and remove `addCallback` since it runs the lambda inline anyways, and triggers unnecessary CUDA stream logic. If we want CUDA operations such as NCCL kernels accurately profiled, we should use the profiler with `use_cuda=True`. However, we are currently debugging a deadlock for the `use_cuda=True` case, fix is being tracked in https://github.com/pytorch/pytorch/issues/48987. 

To ensure that the tests are no longer flaky, submitted this PR to ci-all: https://github.com/pytorch/pytorch/pull/48947 and ran the test a bunch of times ssh'd into the CI machine and made sure they don't fail.

Differential Revision: [D25368322](https://our.internmc.facebook.com/intern/diff/D25368322/)

[ghstack-poisoned]

mrzzd reviewed (approved):
Looks good. Thanks!

rohan-varma committed:
Update on "Fix ProcessGroupNCCL profiling when profiler is not run with use_cuda"


Move `recordFunctionEndCallback` to after the blocking portion of launching the NCCL kernel, and remove `addCallback` since it runs the lambda inline anyways, and triggers unnecessary CUDA stream logic. If we want CUDA operations such as NCCL kernels accurately profiled, we should use the profiler with `use_cuda=True`. However, we are currently debugging a deadlock for the `use_cuda=True` case, fix is being tracked in https://github.com/pytorch/pytorch/issues/48987. 

To ensure that the tests are no longer flaky, submitted this PR to ci-all: https://github.com/pytorch/pytorch/pull/48947 and ran the test a bunch of times ssh'd into the CI machine and made sure they don't fail.

Differential Revision: [D25368322](https://our.internmc.facebook.com/intern/diff/D25368322/)

[ghstack-poisoned]

rohan-varma commented:
Tested on the ci-all PR by ssh'ing into the multigpu instance and running the following:
```
while [ $? -eq 0 ] ; do TEMP_DIR="/tmp" BACKEND="nccl" WORLD_SIZE="3" python test/distributed/test_distributed_fork.py -v TestDistBackendWithFork.test_all_reduce_sum_cuda ; done
```

for test_all_reduce_sum_cuda, test_reduce_multigpu, test_all_gather_multigpu, and test_all_reduce_sum_cuda_async. The command never exited so the tests should no longer be flaky.

rohan-varma committed:
Update on "Fix ProcessGroupNCCL profiling when profiler is not run with use_cuda"


Move `recordFunctionEndCallback` to after the blocking portion of launching the NCCL kernel, and remove `addCallback` since it runs the lambda inline anyways, and triggers unnecessary CUDA stream logic. If we want CUDA operations such as NCCL kernels accurately profiled, we should use the profiler with `use_cuda=True`. However, we are currently debugging a deadlock for the `use_cuda=True` case, fix is being tracked in https://github.com/pytorch/pytorch/issues/48987. 

To ensure that the tests are no longer flaky, submitted this PR to ci-all: https://github.com/pytorch/pytorch/pull/48947 and ran the test a bunch of times ssh'd into the CI machine and made sure they don't fail.

Differential Revision: [D25368322](https://our.internmc.facebook.com/intern/diff/D25368322/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Fix ProcessGroupNCCL profiling when profiler is not run with use_cuda"


Move `recordFunctionEndCallback` to after the blocking portion of launching the NCCL kernel, and remove `addCallback` since it runs the lambda inline anyways, and triggers unnecessary CUDA stream logic. If we want CUDA operations such as NCCL kernels accurately profiled, we should use the profiler with `use_cuda=True`. However, we are currently debugging a deadlock for the `use_cuda=True` case, fix is being tracked in https://github.com/pytorch/pytorch/issues/48987. 

To ensure that the tests are no longer flaky, submitted this PR to ci-all: https://github.com/pytorch/pytorch/pull/48947 and ran the test a bunch of times ssh'd into the CI machine and made sure they don't fail.

Differential Revision: [D25368322](https://our.internmc.facebook.com/intern/diff/D25368322/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Fix ProcessGroupNCCL profiling when profiler is not run with use_cuda"


Move `recordFunctionEndCallback` to after the blocking portion of launching the NCCL kernel, and remove `addCallback` since it runs the lambda inline anyways, and triggers unnecessary CUDA stream logic. If we want CUDA operations such as NCCL kernels accurately profiled, we should use the profiler with `use_cuda=True`. However, we are currently debugging a deadlock for the `use_cuda=True` case, fix is being tracked in https://github.com/pytorch/pytorch/issues/48987. 

To ensure that the tests are no longer flaky, submitted this PR to ci-all: https://github.com/pytorch/pytorch/pull/48947 and ran the test a bunch of times ssh'd into the CI machine and made sure they don't fail.

Differential Revision: [D25368322](https://our.internmc.facebook.com/intern/diff/D25368322/)

[ghstack-poisoned]

rohan-varma commented:
Hm, the bazel build CI job is marked as pending but when I go to it it looks completed - https://app.circleci.com/pipelines/github/pytorch/pytorch/248979/workflows/b86fa46c-c327-4797-991f-562f0ff741cf/jobs/9508765
